adoc/code/specialization_id.cpp

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2011-2021 The Khronos Group, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <sycl/sycl.hpp>
+using namespace sycl; // (optional) avoids need for "sycl::" before SYCL names
+
+struct Mixed {int i; float f;};
+
+constexpr specialization_id<int> a{1};            // OK
+static constexpr specialization_id<int> b{2};     // OK
+specialization_id<int> c{3};                      // ERROR: not constexpr
+
+namespace foo {
+  constexpr specialization_id<Mixed> c{1, 3.14};  // OK
+}
+namespace {
+  constexpr specialization_id<Mixed> d{2, 2.718}; // OK
+}
+
```

Comment:
I have the feeling that
```c++
namespace toto {
  constexpr specialization_id<Mixed> d{3, 4};
}
```
can work since the first `d` can be moved outside of the integration headers while the second one remains as `toto::d`.

Reply:
Have we resolved `constexpr` vs. `inline constexpr`?  I'd rather we just mark them `inline constexpr` when they are non-`static` or we can have examples for both.

Reply:
@keryell: Are you saying that the application could have two `specialization_id` variables with the same name that are declared in different named namespaces?  Yes, I agree that is legal.  I can add an example to demonstrate this.

@nliber: We identified an implementation that allows `specialization_id` variables to be declared either `inline constepxr` or just `constexpr`.  I can add an example showing the `inline constexpr` case.

Reply:
I added some more examples in 95154eb0cbcd03171e9adcd2601f491cb8f45864.

I also realized that the restriction on declaring a `specialization_id` variable in class scope could be simplified.  The description before said that it was OK to declare a `specialization_id` variable in class scope so long as the class was not declared in function scope.  However, I realize now that normal C++ rules forbid a `constexpr` variable from being declared in class scope if the class is in function scope.  Therefore, we can simplify the restriction about `specialization_id` variables to just say that it is OK to declare them in class scope.

Reply:
Responding to @keryell's original question in this thread ... we identified an implementation which allows a `specialization_id` variable to be "shadowed" by an enclosed unnamed namespace, so we removed this restriction from the spec.

Unfortunately, we identified a different shadowing case with `inline` namespaces, so we added a restriction for that.  However, we think this `inline` shadowing is very weird, so we don't think it will be much of a burden.